File: compute-go/translator/expiry.go
package translator

import (
	"strings"
	"time"
)

func translateExpire(args []string) (Translation, error) {
	return translateExpireWith(args, "expire", false)
}

func translateExpireAt(args []string) (Translation, error) {
	return translateExpireWith(args, "expireat", true)
}

func translateExpireWith(args []string, cmd string, absolute bool) (Translation, error) {
	if len(args) < 2 || len(args) > 3 {
		return Translation{}, errWrongArgs(cmd)
	}

	key := args[0]
	value, ok := parseInt64(args[1])
	if !ok {
		return Translation{}, ErrNotInteger
	}

	var expiresAt int64
	if absolute {
		expiresAt = value
	} else {
		expiresAt = time.Now().Unix() + value
	}

	option := ""
	if len(args) == 3 {
		option = strings.ToUpper(args[2])
	}

	updateSQL := "UPDATE keys SET expires_at = ? WHERE key = ? RETURNING key"
	params := []any{expiresAt, key}

	switch option {
	case "":
		// default
	case "NX":
		updateSQL = "UPDATE keys SET expires_at = ? WHERE key = ? AND expires_at IS NULL RETURNING key"
	case "XX":
		updateSQL = "UPDATE keys SET expires_at = ? WHERE key = ? AND expires_at IS NOT NULL RETURNING key"
	case "GT":
		updateSQL = "UPDATE keys SET expires_at = ? WHERE key = ? AND (expires_at IS NULL OR expires_at < ?) RETURNING key"
		params = []any{expiresAt, key, expiresAt}
	case "LT":
		updateSQL = "UPDATE keys SET expires_at = ? WHERE key = ? AND expires_at IS NOT NULL AND expires_at > ? RETURNING key"
		params = []any{expiresAt, key, expiresAt}
	default:
		return Translation{}, errWrongArgs(cmd)
	}

	stmts := []Statement{
		deleteIfExpired(key),
		{
			SQL:    "SELECT expires_at FROM keys WHERE key = ?",
			Params: []any{key},
		},
		{
			SQL:    updateSQL,
			Params: params,
		},
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			if len(results[1]) == 0 {
				return int64(0), nil
			}
			if len(results[2]) == 0 {
				return int64(0), nil
			}
			return int64(1), nil
		},
	}, nil
}

func translateTTL(args []string) (Translation, error) {
	return translateTTLWith(args, "ttl", false)
}

func translatePTTL(args []string) (Translation, error) {
	return translateTTLWith(args, "pttl", true)
}

func translateTTLWith(args []string, cmd string, millis bool) (Translation, error) {
	if len(args) != 1 {
		return Translation{}, errWrongArgs(cmd)
	}

	key := args[0]
	stmts := []Statement{
		deleteIfExpired(key),
		{
			SQL:    "SELECT expires_at, unixepoch() as now FROM keys WHERE key = ?",
			Params: []any{key},
		},
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			if len(results[1]) == 0 {
				return int64(-2), nil
			}
			row := results[1][0]
			expiresValue, ok := row["expires_at"]
			if !ok || expiresValue == nil {
				return int64(-1), nil
			}
			expiresAt, ok := parseInt64(rowString(row, "expires_at"))
			if !ok {
				return nil, ErrNotInteger
			}
			now, ok := parseInt64(rowString(row, "now"))
			if !ok {
				return nil, ErrNotInteger
			}
			ttl := expiresAt - now
			if millis {
				return ttl * 1000, nil
			}
			return ttl, nil
		},
	}, nil
}

func translatePersist(args []string) (Translation, error) {
	if len(args) != 1 {
		return Translation{}, errWrongArgs("persist")
	}

	key := args[0]
	stmts := []Statement{
		deleteIfExpired(key),
		{
			SQL:    "UPDATE keys SET expires_at = NULL WHERE key = ? AND expires_at IS NOT NULL RETURNING key",
			Params: []any{key},
		},
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			if len(results[1]) == 0 {
				return int64(0), nil
			}
			return int64(1), nil
		},
	}, nil
}

File: meowdis/src/translator/expiry.ts
import type { Statement, Translation } from "./types";
import { ErrNotInteger, errWrongArgs } from "./errors";
import { deleteIfExpired, parseIntStrict, rowString } from "./helpers";

export function translateExpireWith(args: string[], cmd: string, absolute: boolean): Translation {
  if (args.length < 2 || args.length > 3) {
    throw errWrongArgs(cmd);
  }

  const key = args[0];
  const value = parseIntStrict(args[1]);
  if (value === null) {
    throw ErrNotInteger;
  }

  const expiresAt = absolute ? value : Math.floor(Date.now() / 1000) + value;
  const option = args.length === 3 ? args[2].toUpperCase() : "";

  let updateSQL = "UPDATE keys SET expires_at = ? WHERE key = ? RETURNING key";
  let params: unknown[] = [expiresAt, key];

  switch (option) {
    case "":
      break;
    case "NX":
      updateSQL =
        "UPDATE keys SET expires_at = ? WHERE key = ? AND expires_at IS NULL RETURNING key";
      break;
    case "XX":
      updateSQL =
        "UPDATE keys SET expires_at = ? WHERE key = ? AND expires_at IS NOT NULL RETURNING key";
      break;
    case "GT":
      updateSQL =
        "UPDATE keys SET expires_at = ? WHERE key = ? AND (expires_at IS NULL OR expires_at < ?) RETURNING key";
      params = [expiresAt, key, expiresAt];
      break;
    case "LT":
      updateSQL =
        "UPDATE keys SET expires_at = ? WHERE key = ? AND expires_at IS NOT NULL AND expires_at > ? RETURNING key";
      params = [expiresAt, key, expiresAt];
      break;
    default:
      throw errWrongArgs(cmd);
  }

  const statements: Statement[] = [
    deleteIfExpired(key),
    { sql: "SELECT expires_at FROM keys WHERE key = ?", params: [key] },
    { sql: updateSQL, params },
  ];

  return {
    statements,
    mapResult: (results) => {
      if (results[1].length === 0) {
        return 0;
      }
      if (results[2].length === 0) {
        return 0;
      }
      return 1;
    },
  };
}

export function translateExpire(args: string[]): Translation {
  return translateExpireWith(args, "expire", false);
}

export function translateExpireAt(args: string[]): Translation {
  return translateExpireWith(args, "expireat", true);
}

export function translateTTLWith(args: string[], cmd: string, millis: boolean): Translation {
  if (args.length !== 1) {
    throw errWrongArgs(cmd);
  }

  const key = args[0];
  const statements: Statement[] = [
    deleteIfExpired(key),
    { sql: "SELECT expires_at, unixepoch() as now FROM keys WHERE key = ?", params: [key] },
  ];

  return {
    statements,
    mapResult: (results) => {
      if (results[1].length === 0) {
        return -2;
      }
      const row = results[1][0];
      const expiresValue = row["expires_at"];
      if (expiresValue === null || expiresValue === undefined) {
        return -1;
      }
      const expiresAt = parseIntStrict(rowString(row, "expires_at"));
      const now = parseIntStrict(rowString(row, "now"));
      if (expiresAt === null || now === null) {
        throw ErrNotInteger;
      }
      const ttl = expiresAt - now;
      return millis ? ttl * 1000 : ttl;
    },
  };
}

export function translateTTL(args: string[]): Translation {
  return translateTTLWith(args, "ttl", false);
}

export function translatePTTL(args: string[]): Translation {
  return translateTTLWith(args, "pttl", true);
}

export function translatePersist(args: string[]): Translation {
  if (args.length !== 1) {
    throw errWrongArgs("persist");
  }

  const key = args[0];
  const statements: Statement[] = [
    deleteIfExpired(key),
    {
      sql: "UPDATE keys SET expires_at = NULL WHERE key = ? AND expires_at IS NOT NULL RETURNING key",
      params: [key],
    },
  ];

  return {
    statements,
    mapResult: (results) => (results[1].length === 0 ? 0 : 1),
  };
}
